lib/hooks/userspace/hook-koa.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const cls = require('../../cls.js');
+const TraceLabels = require('../../trace-labels.js');
+const shimmer = require('shimmer');
+const semver = require('semver');
+const constants = require('../../constants.js');
+var agent;
+
+const SUPPORTED_VERSIONS = '1.x';
+
+function useWrap(use) {
+  return function useTrace() {
+    if (!this._google_trace_patched) {
+      this._google_trace_patched = true;
+      this.use(middleware);
+    }
+    return use.apply(this, arguments);
+  };
+}
+
+function* middleware(next) {
+  /* jshint validthis:true */
+  const namespace = cls.getNamespace();
+  if (!namespace) {
+    agent.logger.info('Koa: no namespace found, ignoring request');
+    return;
+  }
+  if (!agent.shouldTrace(this.req.url)) {
+    return;
+  }
+  const req = this.req;
+  const res = this.res;
+
+  namespace.bindEmitter(req);
+  namespace.bindEmitter(res);
+
+  const originalEnd = res.end;
+
+  namespace.run(function() {
+    const rootContext = startRootSpanForRequest(req);
+
+    // wrap end
+    res.end = function(chunk, encoding) {
+      res.end = originalEnd;
+      const returned = res.end(chunk, encoding);
+
+      endRootSpanForRequest(rootContext, req, res);
+      return returned;
+    };
+    namespace.bind(next);
+  });
+  yield next;
+}
```

Comment:
The issue is that we would need to wrap the body of that middleware in a `namespace.run` however we would be unable to yield out of the middle of the run function. What we would need is something like:

```javascript
app.use(function *(next){
  const namespace = cls.getNamespace();
  const that = this;
  const rootContext = startRootSpanForRequest(that.req);
  namespace.run(function() {
    yield_from_outer_function next;
    endRootSpanForRequest(rootContext, that.req, that.res);
  });
});
```

However I don't know of an operation providing the `yield_from_outer_function` functionality.